File: src/rust/planning_poker/src/domains/invitation.rs
use super::game::GameId;

#[cfg(test)]
mod tests;

#[derive(PartialEq)]
pub struct InvitationSignature(String);

#[derive(PartialEq)]
pub struct Invitation {
    game_id: GameId,
    signature: InvitationSignature,
}

impl InvitationSignature {
    pub fn new(game_id: GameId) -> Self {
        let hash = sha256::digest(game_id.to_string());
        InvitationSignature(hash)
    }
}

impl ToString for InvitationSignature {
    fn to_string(&self) -> String {
        self.0.to_string()
    }
}

impl Invitation {
    pub fn new(game_id: GameId) -> Self {
        let signature = InvitationSignature::new(game_id);
        Self { game_id, signature }
    }

    pub fn game_id(&self) -> &GameId {
        &self.game_id
    }

    pub fn signature(&self) -> &InvitationSignature {
        &self.signature
    }
}

File: src/domains/invitation.ts
import { Id } from "./game";
import sha256 from "crypto-js/sha256";
import { Branded } from "./type";

const tag = Symbol("InvitationSignature");
export type InvitationSignature = Branded<string, typeof tag>;

export interface T {
  readonly gameId: Id;
  readonly signature: InvitationSignature;
}

export const create = (gameId: Id): T => {
  const hash = sha256(gameId);
  return {
    gameId,
    signature: hash.toString() as InvitationSignature,
  };
};
